src/threading/TaskScheduler.cpp

```diff
@@ -18,62 +19,231 @@
+        /* mutex scope */ {
+            const std::lock_guard<std::mutex> queue_lock(queue_mutex);
+            queue_condition.notify_all();
+        }
+
+        // Now wait for all the threads to finish executing
+        for (auto& thread : threads) {
+            try {
+                if (thread->joinable()) {
+                    thread->join();
+                }
+            }
+            // This gets thrown some time if between checking if joinable and joining
+            // the thread is no longer joinable
+            catch (const std::system_error&) {
+            }
+        }
+    }
+
+    void TaskScheduler::shutdown() {
+        started.store(false);
+        running.store(false);
+        const std::lock_guard<std::mutex> queue_lock(queue_mutex);
+        queue_condition.notify_all();
     }
 
     void TaskScheduler::submit(std::unique_ptr<ReactionTask>&& task) {
 
-        // We do not accept new tasks once we are shutdown
-        if (running) {
+        // Extract the thread pool descriptor from the current task
+        const util::ThreadPoolDescriptor current_pool = task->thread_pool_descriptor;
+
+        // Make sure the pool is created
+        create_pool(current_pool);
+
+        // Make sure we know about this group
+        /* mutex scope */ {
+            const std::lock_guard<std::mutex> group_lock(queue_mutex);
+            const uint64_t group_id = task->group_descriptor.group_id;
+            if (groups.count(group_id) == 0) {
+                groups.emplace(group_id, 0);
+            }
+        }
+
+        // Check to see if this task was the result of `emit<Direct>`
+        // Direct tasks can run after shutdown and before starting, provided they can be run immediately
+        if (task->immediate) {
+            // Map the current thread to the thread pool it belongs to
+            uint64_t thread_pool = util::ThreadPoolDescriptor::DEFAULT_THREAD_POOL_ID;
+            /* mutex scope */ {
+                const std::lock_guard<std::mutex> pool_lock(pool_mutex);
+                if (pool_map.count(std::this_thread::get_id()) > 0) {
+                    thread_pool = pool_map.at(std::this_thread::get_id());
+                }
+            }
 
-            /* Mutex Scope */ {
-                const std::lock_guard<std::mutex> lock(mutex);
-                queue.push(std::forward<std::unique_ptr<ReactionTask>>(task));
+            // Check to see if this task is runnable in the current thread
+            // If it isn't we can just queue it up with all of the other non-immediate task
+            if (is_runnable(task, thread_pool)) {
+                run_task(std::move(task));
+                return;
             }
         }
 
-        // Notify a thread that it can proceed
-        condition.notify_one();
+        // We do not accept new tasks once we are shutdown
+        if (running.load()) {
+            const std::lock_guard<std::mutex> queue_lock(queue_mutex);
```

Comment:
May as well just acquire this lock one line up as we just acquire it again as soon as we leave the if